### DISCUSSION TITLE
A standalone header to interface with the Objeck VM

### DISCUSSION THREAD

**Discussion by 1997tn** [objeck/objeck-lang#110] [Ideas]
Like JNI of Java. You do not need the Java source code. The only thing you will need is the standalone headers on the `include` directory of your JRE and of course the JRE itself. You can even embed jvm.dll into your application. You can do all of that without Java source code.

From the header `lib_api.h` of Objeck, it included `common.h` and `common.h` including a bunch of other headers including external libraries like OpenSSL. And the fact is there is no `Objeck.dll` but `common.cpp` together with `common.h` this indicated that you will need the Objeck source code to do the same thing you could do with JNI but without Java source code.

I know Objeck is open source and Java has been designed this way could be because it used to be closed source. But the way Java doing it is worth to learn from. The whole Objeck source I cloned from this repo is large. Instead if I could just include the headers shipped as part of an Objeck binary distribution like Java and link with `Objeck.dll` thing will be much simple and elegant (and easier too).

**Comment by objeck** [comment]
Thanks for the note, is the ask to have one library to link against for building native C++ modules?

**Reply by 1997tn** [reply]
I give idea but do not ask for it. I think it's a good idea that will improve the users' experience with Objeck though.

**Reply by 1997tn** [reply]
I'm very bad at describing, you should download the JRE and see how they do it.

**Reply by objeck** [reply]
Thanks for the note. The existing Objeck C++ API was not designed to be "open"; thus, it is a collection of help functions. Time permitting, I will explore ways to expose an easier-to-use extension framework.

**Comment by objeck** [comment]
Please take a look at the [following](https://github.com/objeck/objeck-lang/tree/master/core/module).
